infobind/cache.go

```
@@ -0,0 +1,215 @@
+}
+
+// bindCache hold a bindDataMap, key:origin sql hash value: bindData slice.
+type bindCache struct {
+	Cache map[string][]*bindData
+}
+
+// Handle hold a atomic bindCache.
+type Handle struct {
+	bind atomic.Value
+}
+
+// HandleUpdater use to update the bindCache.
+type HandleUpdater struct {
+	parser         *parser.Parser
+	lastUpdateTime types.Time
+	ctx            sessionctx.Context
+	globalHandle   *Handle
+}
+
+type bindRecord struct {
+	OriginalSQL string
+	BindSQL     string
+	Db          string
+	Status      int64
+	CreateTime  types.Time
+	UpdateTime  types.Time
+	Charset     string
+	Collation   string
+}
+
+// NewHandleUpdater create a new HandleUpdater.
+func NewHandleUpdater(handle *Handle, parser *parser.Parser, ctx sessionctx.Context) *HandleUpdater {
+	return &HandleUpdater{
+		globalHandle: handle,
+		parser:       parser,
+		ctx:          ctx,
+	}
+}
+
+// NewHandle create a Handle with a bindCache.
+func NewHandle() *Handle {
+	handle := &Handle{}
+	return handle
+}
+
+// Get get bindCache from a Handle.
+func (h *Handle) Get() *bindCache {
+	bc := h.bind.Load()
+
+	if bc != nil {
+		return bc.(*bindCache)
+	}
+
+	return &bindCache{
+		Cache: make(map[string][]*bindData, defaultBindCacheSize),
+	}
+}
+
+// LoadDiff use to load new bind info to bindCache bc.
+func (h *HandleUpdater) loadDiff(sql string, bc *bindCache) error {
+	recordSets, err := h.ctx.(sqlexec.SQLExecutor).Execute(context.Background(), sql)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	rs := recordSets[0]
+	defer terror.Call(rs.Close)
+
+	chkBatch := rs.NewRecordBatch()
+	for {
+		err = rs.Next(context.TODO(), chkBatch)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		if chkBatch.NumRows() == 0 {
+			return nil
+		}
+		it := chunk.NewIterator4Chunk(chkBatch.Chunk)
+		for row := it.Begin(); row != it.End(); row = it.Next() {
+			record := decodeBindTableRow(row)
+			err = bc.appendNode(h.ctx, record, h.parser)
+			if err != nil {
+				continue
```

Why do we ignore the error here?

---

Should we add a metrics for this error?

---

1. my fault, we can't ignore the error, should throw the error to caller.
2. caller will log the error